### PR TITLE
[Merged by Bors] - feat: define `AddConstEquiv`

### PR DESCRIPTION
**Pull request by urkud**
Also revert some `scoped` modifiers on `simp`s.

---

Waiting for a benchmark.

- [x] depends on: #9725
- [x] depends on: #9724

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/AddConstMap/Equiv.lean [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+open scoped AddConstMap
+
+/-- An equivalence between `G` and `H` conjugating `(· + a)` to `(· + b)`. -/
+structure AddConstEquiv (G H : Type*) [Add G] [Add H] (a : G) (b : H)
+  extends G ≃ H, G →+c[a, b] H
+
+/-- Interpret an `AddConstEquiv` as an `Equiv`. -/
+add_decl_doc AddConstEquiv.toEquiv
+
+/-- Interpret an `AddConstEquiv` as an `AddConstMap. -/
```
Comment (ericrbg) [inline Mathlib/Algebra/AddConstMap/Equiv.lean:26]:
```suggestion
/-- Interpret an `AddConstEquiv` as an `AddConstMap.` -/
```

File: Mathlib/Algebra/AddConstMap/Equiv.lean [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  coe_injective' | ⟨_, _⟩, ⟨_, _⟩, h, _ => by congr; exact DFunLike.ext' h
+  map_add_const f x := f.map_add_const' x
+
+namespace AddConstEquiv
+
+variable {G H K : Type*} [Add G] [Add H] [Add K] {a : G} {b : H} {c : K}
+
+@[ext] lemma ext {e₁ e₂ : G ≃+c[a, b] H} (h : ∀ x, e₁ x = e₂ x) : e₁ = e₂ := DFunLike.ext _ _ h
+
+lemma toEquiv_injective : Injective (toEquiv : (G ≃+c[a, b] H) → G ≃ H) :=
```
Comment (ericrbg) [inline Mathlib/Algebra/AddConstMap/Equiv.lean:61]: should this not be done by some general `EquivLike` plumbing?

**Comment by eric-wieser** [comment]
Should this be written in terms of `VAdd` instead? Then you could show that all affine equivalences are `VAddConstEquiv`s

File: Mathlib/Algebra/AddConstMap/Equiv.lean (hunk 2 above)
Comment (urkud) [reply Mathlib/Algebra/AddConstMap/Equiv.lean]: I see that other `*Equiv` prove `toEquiv_injective` first, then use it for `EquivLike`. I changed this file to follow the same pattern. We may change it everywhere later.

**Comment by urkud** [comment]
@eric-wieser The only application I have in mind is a lift of a circle map/homeomorphism to the real line. Do you have some applications of `VAdd` version in mind? I mean, some nontrivial result that will benefit from this generalization. If no, then I would prefer to postpone this generalization until we need it.

**Comment by urkud** [comment]
!bench

File: Mathlib/Algebra/AddConstMap/Equiv.lean [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  coe_injective' _ _ h _ := toEquiv_injective <| DFunLike.ext' h
+
+instance {G H : Type*} [Add G] [Add H] {a : G} {b : H} :
+    AddConstMapClass (G ≃+c[a, b] H) G H a b where
+  map_add_const f x := f.map_add_const' x
+
+@[ext] lemma ext {e₁ e₂ : G ≃+c[a, b] H} (h : ∀ x, e₁ x = e₂ x) : e₁ = e₂ := DFunLike.ext _ _ h
+
+@[simp]
+lemma toEquiv_eq_iff {e₁ e₂ : G ≃+c[a, b] H} : e₁.toEquiv = e₂.toEquiv ↔ e₁ = e₂ :=
```
Comment (YaelDillies) [inline Mathlib/Algebra/AddConstMap/Equiv.lean:55]:
```suggestion
lemma toEquiv_inj {e₁ e₂ : G ≃+c[a, b] H} : e₁.toEquiv = e₂.toEquiv ↔ e₁ = e₂ :=
```

File: Mathlib/Algebra/AddConstMap/Basic.lean [hunk 4]
```diff
@@ -383,15 +383,22 @@ instance : Monoid (G →+c[a, a] G) :=
   DFunLike.coe_injective.monoid (M₂ := Function.End G) _ rfl (fun _ _ ↦ rfl) fun _ _ ↦ rfl
 
 theorem mul_def (f g : G →+c[a, a] G) : f * g = f.comp g := rfl
-@[scoped simp] theorem coe_mul (f g : G →+c[a, a] G) : ⇑(f * g) = f ∘ g := rfl
+@[simp] theorem coe_mul (f g : G →+c[a, a] G) : ⇑(f * g) = f ∘ g := rfl
```
[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Algebra/AddConstMap/Basic.lean:386]: Will it `norm_cast` RHS to LHS? I prefer to `norm_cast` it to `.comp`. Do we have "`push_cast` but no `norm_cast`" lemmas?

[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/AddConstMap/Basic.lean:386]: Yes that's one of the settings to norm_cast, although I don't remember how  it's called so you'll need to look through the docs

[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Algebra/AddConstMap/Basic.lean]: There is a `push_cast` simp attribute. Added.

**Comment by YaelDillies** [comment]
Let's see whether it now works from my phone

maintainer merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**